<a href="https://colab.research.google.com/github/manluz555-ops/Line_progr/blob/main/HW_8_Manzar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашня робота № 8
Манзар Л.В.

In [2]:
# 1. Завантаження набору даних Iris
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Завантажуємо дані
iris = load_iris()
X = iris.data
y = iris.target
print("Форма даних:", X.shape)
print("Класи:", np.unique(y))

# 2. Розподіл на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)
print("Train:", X_train.shape, "Test:", X_test.shape)

# 3. Вибірка ознак окремо для кожного класу
classes = np.unique(y_train)
class_data = {c: X_train[y_train == c] for c in classes}

# 4. Матриці коваріації
cov_matrices = {c: np.cov(class_data[c].T) for c in classes}

# 5. Обернені матриці коваріації
inv_cov_matrices = {c: np.linalg.inv(cov_matrices[c]) for c in classes}

# 6. Апріорні ймовірності
priors = {c: len(class_data[c]) / len(X_train) for c in classes}
print("Апріорні ймовірності:", priors)

# 7. Функція дискримінантної функції для одного вектора
def discriminant_function(x, mean, inv_cov, prior):
    diff = x - mean
    return -0.5 * diff @ inv_cov @ diff.T + np.log(prior)

# 8. Функція для всієї матриці тестових даних
def predict_custom(X_test):
    predictions = []
    for x in X_test:
        scores = {}
        for c in classes:
            mean = np.mean(class_data[c], axis=0)
            scores[c] = discriminant_function(x, mean, inv_cov_matrices[c], priors[c])
        predictions.append(max(scores, key=scores.get))
    return np.array(predictions)

y_pred_custom = predict_custom(X_test)

# 9. Прогнозування через QuadraticDiscriminantAnalysis
qda = QuadraticDiscriminantAnalysis()
qda.fit(X_train, y_train)
y_pred_qda = qda.predict(X_test)

# 10. Порівняння результатів
accuracy_custom = np.mean(y_pred_custom == y_test)
accuracy_qda = np.mean(y_pred_qda == y_test)

print("Точність власної реалізації:", accuracy_custom)
print("Точність sklearn QDA:", accuracy_qda)



Форма даних: (150, 4)
Класи: [0 1 2]
Train: (105, 4) Test: (45, 4)
Апріорні ймовірності: {np.int64(0): 0.3333333333333333, np.int64(1): 0.3333333333333333, np.int64(2): 0.3333333333333333}
Точність власної реалізації: 0.9777777777777777
Точність sklearn QDA: 0.9777777777777777


## Висновки

1. Було завантажено класичний набір даних **Iris**, що містить 150 спостережень із трьома класами (setosa, versicolor, virginica). Дані розподілено на навчальну та тестову вибірки у співвідношенні 70/30.

2. Для кожного класу було обчислено матриці коваріації та їх обернені значення. Це дозволило реалізувати дискримінантні функції у рамках квадратичного дискримінантного аналізу.

3. Апріорні ймовірності класів визначені на основі частки кожного класу у тренувальних даних, що забезпечує коректне врахування їхнього розподілу.

4. Реалізовано власну функцію дискримінантного аналізу, яка для кожного вектора тестових даних обчислює значення дискримінантних функцій та визначає клас із найбільшою ймовірністю.

5. Порівняння результатів показало, що точність власної реалізації є близькою до точності стандартної функції **QuadraticDiscriminantAnalysis** із бібліотеки sklearn. Це підтверджує правильність реалізації алгоритму.

### Загальний підсумок
Задача продемонструвала практичну реалізацію квадратичного дискримінантного аналізу: від обчислення матриць коваріації до прогнозування класів. Власна реалізація показала результати, співставні з бібліотечною функцією sklearn, що свідчить про коректність алгоритму та його застосування для класифікації багатовимірних даних.
